ipd/0020/README.adoc

```diff
@@ -0,0 +1,770 @@
+While this is easy for some APIs (e.g., kmem), it is probably much
+harder for others. Furthermore, it requires duplicate code both in
+terms of the DDI/DKI as well as the module-under-test. Not to mention
+other potential booby traps like compilation differences.
+
+The ktest facility avoids all of this by bringing the tests to the
+kernel. Everything is there already, no faking required. Furthermore,
+in some ways, funnily enough, debugging via DTrace is easier once in
+the kernel (fbt vs. pid), and if you have a complicated test that
+needs deep debugging you could always use DTrace to force a panic of
+the kernel on a specific test failure in order to perform post-mortem
+debugging with mdb.
+
+=== simnet
+
+The simnet device provides a pseudo mac device (also known as a mac
+provider). This is a device that implemens the mac(9E) interface but
+is purely virtual and allows user configuration via the `dladm(1M)`
+command. This is a very powerful device when combined with bridges, IP
+routing, and zones, because it allows full emulating of an arbitrary
+network on one host. However, this is obviously a very specialized
+form of testing. It is complimentary to ktest. Unfortunately we
+currently don't document simnet, but you can find out more at my blog
+[<<resurrect-simnet>>] [<<simnet-basics>>].
+
+=== Internet Packet Disturber (IPD)
+
+The internet packet disturber (or `ipd` for short) is a little known
+tool created by Robert Mustacchi. It is used to simlate congested and
+lossy networks where they don't actually exist. This allows one to
+test how upper layer connection-based protocols, like TCP, handle a
+lossy netowrk. Useful for testing say TCP congestion algorithms and
+retransmit behavior. It's also useful to see how any application-layer
+protocols react to such a network. Once again, this is a specialized
+testing tool which is complimentary to ktest.
+
+To find out more see Robert's lovely big-theory statement on ipd
+[<<ipd-theory>>] and see the ipdadm(1M) man page[<<ipdadm>>].
+
+=== pshot: Psuedo Bus Nexus Driver
+
+This is a pseudo device that allows one to create an arbitrarily
+complex device tree. It looks like this tool was created by Garrett
+D'Amore and provides something similar, in spirit, to simnet, but
+instead targets PCI devices. Once again, this feels like a
+complimentary tool.
+
+=== Miscellaneous
+
+It seems there are several other miscellaneous test drivers, such as
+`gen_drv` (Generic Character Device) and `emul64`, which I did not dig
+further into. In fact, it appears there is a package called
+`/system/io/tests` that consolidates many of these drivers, including
+the aforementioned pshot. If someone wants to give me the skinny on
+this package and its drivers I'd love to know more. That said, I don't
+think any of these things overlap with ktest, and I also don't think
+ktest should be delivered as part of this package. Rather, I think it
+should have its own.
+
```

Comment:
Among "related work" you could mention the crypto tests:
usr/src/test/crypto-tests/tests/common/cryptotest_kcf.c
that came in with 5869 Need AES CMAC support in KCF+PKCS11
Those are conceptually similar to what you're proposing here,
specific to the KCF functions.

Reply:
Awesome, I will look at that and probably add a blurb about it.